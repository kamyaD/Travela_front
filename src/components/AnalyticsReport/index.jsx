import React, { Component, Fragment } from 'react';
import { PropTypes } from 'prop-types';
import './index.scss';
import errorIcon from '../../images/error_24px.svg';
import download from '../../images/icons/save_alt_24px.svg';
import TravelReadiness   from '../TravelReadiness';

export default class AnalyticsReport extends Component {

  componentDidMount() {
    const { fetchDepartmentTrips, fetchReadiness } = this.props;
    fetchDepartmentTrips({filterBy: 'month', type: 'json'});
    fetchReadiness({page: '1', limit: '6', type:'json'});
  }

  getDepartmentTripsCSV = () => {
    const { fetchDepartmentTrips } = this.props;
    fetchDepartmentTrips({ filterBy: 'month', type: 'file' });
    fetchDepartmentTrips({ filterBy: 'month', type: 'json' });
  }

  renderButton = (name, icon, text, onclickFunction) => (
    <button
      name={name}
      id={name}
      type="button"
      className="analyticsReport__export-button"
      onClick={onclickFunction}>
      <Fragment>
        {text}
        <img src={icon} alt={text} />
      </Fragment>
    </button>
  );

  renderTripsDetails = (item) => {
    return (
      <div className="analyticsReport__row analyticsReport__report-details" key={item.label}>
        <div>
          <p>{item.value}</p>
        </div>
        <div>
          <p>{item.label}</p>
        </div>
      </div>
    );
  }

  renderNotFound = () => {
    return (
      <div id="no-records" className="analyticsReport__report-details">
        <br />
        <p className="analyticsReport__text-center">No records found</p>
      </div>
    );
  }

  renderSpinner = () => {
    return (
      <div className="analyticsReport__report-details">
        <br />
        <div className="analyticsReport__spinner" />
        <p className="analyticsReport__text-center">generating report...</p>
      </div>
    );
  }
  render() {
    const { departmentTrips, readiness, fetchReadiness, exportReadiness } = this.props;
    const { report, loading } = departmentTrips;
    return(
      <div className="analyticsReport">
        <TravelReadiness
          readiness={readiness}
          renderNotFound={this.renderNotFound}
          renderButton={this.renderButton}
          fetchReadiness={fetchReadiness}
          exportReadiness={exportReadiness}
          renderSpinner={this.renderSpinner} />
        <div className="analyticsReport__card">
          <div className="analyticsReport__row analyticsReport__header">
            <p>Number of Trips/Month</p>
            {this.renderButton('btnExportTripsPerMonth', download, 'Export', this.getDepartmentTripsCSV)}
          </div>
          <div className="analyticsReport__row analyticsReport__report-header">
            <div>
              <p>Numbers</p>
            </div>
            <div>
              <p>Departments</p>
            </div>
          </div>
          {loading &&
            this.renderSpinner()
          }
          {report && report.length > 0 && !loading &&
            report.map(item => this.renderTripsDetails(item))}
          {report && !report.length &&
            this.renderNotFound()
          }
        </div>
      </div>
    );
  }
}

AnalyticsReport.propTypes = {
  departmentTrips: PropTypes.object.isRequired,
  readiness: PropTypes.object.isRequired,
  fetchDepartmentTrips: PropTypes.func.isRequired,
  fetchReadiness:PropTypes.func.isRequired,
  exportReadiness:PropTypes.func.isRequired
};
