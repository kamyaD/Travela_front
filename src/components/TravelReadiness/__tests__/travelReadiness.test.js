import React from 'react';
import { shallow, mount } from 'enzyme';
import {Provider} from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { TravelReadiness } from '../index';
import travelReadinessData from '../__mocks__/travelReadinessMockData';
import AnalyticsReport from '../../AnalyticsReport';

const defaultProps = {
  readiness: {
    response: travelReadinessData,
    isLoading: false
  },
  fetchReadiness: jest.fn(),
  renderButton: jest.fn(),
  departmentTrips: {
    report: ''
  },
  fetchDepartmentTrips: jest.fn,
};
const setup = (props) => {
  const actualProps = props || defaultProps;
  return shallow(<TravelReadiness {...actualProps} />);
};
describe('Test suite for Travel Readiness Component', () => {
  it('should render the component properly', () => {
    const wrapper = setup();
    expect(wrapper.length).toBe(1);
  });
  it('should match the snapshot', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should call fetchReadiness when export button is clicked', () => {
    const mountWrapper = mount(
      <Provider>
        <MemoryRouter>
          <AnalyticsReport {...defaultProps} />
        </MemoryRouter>
      </Provider>
    );
    const button = mountWrapper.find('.analyticsReport__export-button').first();
    button.simulate('click');
    expect(defaultProps.fetchReadiness).toHaveBeenCalled();
  });
});
 
