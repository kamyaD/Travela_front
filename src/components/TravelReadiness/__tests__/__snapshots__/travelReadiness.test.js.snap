// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for Travel Readiness Component should match the snapshot 1`] = `
<div
  className="analyticsReport__card"
  style={
    Object {
      "marginRight": "30px",
    }
  }
>
  <div
    className="analyticsReport__row analyticsReport__header"
  >
    <p>
      Travel Readiness
    </p>
  </div>
  <div
    className="analyticsReport__row analyticsReport__report-header"
  >
    <div>
      <p>
        Name
      </p>
    </div>
    <div>
      <p>
        % Complete
      </p>
    </div>
    <div>
      <p>
        Expected Arrival Date
      </p>
    </div>
  </div>
</div>
`;
